EIPS/eip-2771.md

```diff
@@ -0,0 +1,211 @@
+author: Ronan Sandford (@wighawag), Liraz Siri (@lirazsiri),  Dror Tirosh (@drortirosh), Yoav Weiss (@yoavw), Alex Forshtat (@forshtat), Hadrien Croubois (@Amxx), Tomar Sachin (@tomarsachin2271), Patrick McCorry (@stonecoldpat), Nicolas Venturo (@nventuro), Fabian Vogelsteller (@frozeman) 
+discussions-to: https://github.com/ethereum/EIPs/pull/2771
+status: Draft
+type: Standards Track
+category: ERC
+created: 2020-07-01
+---
+
+## Simple Summary
+
+A contract interface for receiving meta transactions through a trusted
+forwarder.
+
+## Abstract
+
+This ERC defines a minimal contract-level protocol that a compliant Recipient
+contract needs to support in order to be capable of accepting a meta
+transaction through a compliant Forwarder contract that it trusts to help it
+identify the address of the Transaction Signer. 
+
+No EVM-level protocol changes are proposed or required.
+
+## Motivation
+
+There is a growing interest in making it possible for Ethereum contracts to
+accept calls from externally owned accounts that do not have ETH to pay for
+gas. 
+
+This can be accomplished with meta transactions, which are transactions that
+have been:
+
+1. Authorized by the **Transaction Signer**. For example, signed by an
+   externally owned account.
+2. Relayed by an untrusted third party that pays for the gas (the **Gas
+   Relay**)
+
+`msg.sender` is a transaction parameter that can be inspected by a contract to
+determine who signed the transaction. The integrity of this parameter is
+guaranteed by the Ethereum EVM, but for a meta transaction securing
+`msg.sender` is insufficient.
+
+The problem is that for a contract that is not natively aware of meta
+transactions, the `msg.sender` of the transaction will make it appear to be
+coming from the **Gas Relay** and not the **Transaction Signer**. A secure
+protocol for a contract to accept meta transactions needs to prevent the **Gas
+Relay** from forging, modifying or duplicating requests by the **Transaction
+Signer**.
+
+## Specification
+
+The keywords "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
+interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)).
+
+Here is an example flow:
+
+![Example flow](../assets/eip-2771/example-flow.png)
+
+
+*   **Transaction Signer** - entity that signs & sends to request to **Gas
+    Relay**
+*   **Gas Relay** - receives a signed request off-chain from **Transaction
+    Signer** and pays gas to turn it into a valid transaction that goes through
+**Trusted Forwarder**
+*   **Trusted Forwarder** - a contract that is trusted by the `Recipient` to
+    correctly verify the signature and nonce before forwarding the request from
+**Transaction Signer**
+*   **Recipient** - a contract that can securely accept meta-transactions
+    through a **Trusted Forwarder** by being compliant with this standard.
+
```

Comment:
It seems that the specification section does not explicitly state the specification: the sender address is appended to the end of the calldata.

It is only explained under the "extracting" and "example" subsections.

Reply:
"The Trusted Forwarder is responsible for calling the Recipient contract and MUST append the address of the Transaction Signer (20 bytes of data) to the end of the call data."

Is it unclear? Would duplicating the above paragraph or parts of it help? Or are you suggesting to move it somewhere else? The specification section is divided into subsections for clarity and context. 

The rationale for having it below the "Extracting The Transaction Signer address" heading, is that explaining what needs to happen in terms extraction of the signer helps give additional context that may help explain why the signer was appended to the calldata.

Reply:
The `Extracting The Transaction Signer Address` is a subsection of the `Specification` section, so I think it is fine there.

Reply:
You may want to put the above definitions into a `### Definitions` subsection though, just to help with automated table of contents generation.